<a href="https://colab.research.google.com/github/12345997/WEEK-9-REPO/blob/main/week_9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Defining the Question

### a) Specifying the Data Analytic Question

> Predict predict whether a message is spam or not

### b) Defining the Metric for Success
to be able to carry out data analysis and modelling using various algorithm to check the possibility that a message is a spam or not

### c) Understanding the context 
spam are used by companies in dvertising their products and for that resason one needs to filter out whether a message is a spam or not

### d) Recording the Experimental Design

### e) Data Relevance
the dataset we are provided with was relevant for our analysis 

## 2. Reading the Data

In [ ]:
#importing libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.metrics import confusion_matrix
import numpy as np
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.naive_bayes import BernoulliNB
from tabulate import tabulate

In [ ]:
#Read data from a csv to a dataframe
data = pd.read_csv("/content/spambase.data", names=list(range(0,58)))
#data = df.values.tolist()


#Data preparation

In [ ]:
#checking for null values
data.isnull().any()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35    False
36    False
37    False
38    False
39    False
40    False
41    False
42    False
43    False
44    False
45    False
46    False
47    False
48    False
49    False
50    False
51    False
52    False
53    False
54    False
55    False
56    False
57    False
dtype: bool

In [ ]:
data.duplicated().sum()

391

In [ ]:
 data.drop_duplicates(subset=None, inplace=True)

#Exploratory data analysis

## . Implementing the Solution

In [ ]:
X = data.iloc[:, :-1].values
y = data.iloc[:, 57].values

In [ ]:
# after performing feature extraction from our data, it is time to build our model. 
#  start by splitting our data into training and test sets with a 20% test set and 80% as training set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)


In [ ]:
#initializing the naive_bayes therorem
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB().fit(X_train, y_train)


In [ ]:
# Once we have put together our classifier, we can evaluate its performance in the testing set
predicted = model.predict(X_test)
print(np.mean(predicted == y_test))

0.7947882736156352


thus our model gives an predicted accuracy value of 79.47% which is acceptable

In [ ]:
# after performing feature extraction from our data, it is time to build our model. 
#  start by splitting our data into training and test sets with a 20% test set and 80% as training set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)


In [ ]:
#initializing the naive_bayes therorem
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB().fit(X_train, y_train)


In [ ]:
# Once we have put together our classifier, we can evaluate its performance in the testing set
predicted = model.predict(X_test)
print(np.mean(predicted == y_test))

0.7921795800144823


thus our model gives an predicted accuracy value of 79.21% which is acceptable

In [ ]:
# after performing feature extraction from our data, it is time to build our model. 
#  start by splitting our data into training and test sets with a 20% test set and 80% as training set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40)


In [ ]:
#initializing the naive_bayes therorem
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB().fit(X_train, y_train)

In [ ]:
# Once we have put together our classifier, we can evaluate its performance in the testing set
predicted = model.predict(X_test)
print(np.mean(predicted == y_test))

0.7941336230309615


with 40% test set the accuracy of the models predictions is 79.41% which is nearly the saame as the predicted accuracy with 20% test set

## 8. Challenging the solution

In [ ]:
# Reviewing the Solution 
#
the difference between the three models is small hence the model can be said to be worth for our training dataset

## 9. Follow up questions

> At this point, we can refine our question or collect new data, all in an iterative process to get at the truth.



### a). Did we have the right data? **yes**

### b). Do we need other data to answer our question? **yes, so the we increase the accuracy of our prediction**

### c). Did we have the right question? **yes**